src/flake8/main/application.py

```diff
@@ -115,8 +115,7 @@ def parse_preliminary_options(
 
     def exit_code(self) -> int:
         """Return the program exit code."""
-        assert self.options is not None
-        if self.options.exit_zero:
```

Comment:
This is unrelated to your change. Please revert it

Reply:
Without changing the two assertions, the output is:
```console
$ flake8 --config foo.cfg t.py
There was a critical error during execution of Flake8:
The specified config file does not exist: foo.cfg
Traceback (most recent call last):
  File "/home/rkm/dev/pycqa/flake8/venv/bin/flake8", line 33, in <module>
    sys.exit(load_entry_point('flake8', 'console_scripts', 'flake8')())
  File "/home/rkm/dev/pycqa/flake8/src/flake8/main/cli.py", line 22, in main
    app.run(argv)
  File "/home/rkm/dev/pycqa/flake8/src/flake8/main/application.py", line 395, in run
    assert self.options is not None
AssertionError
```
Compared to simply:
```console
$ flake8 --config foo.cfg t.py
There was a critical error during execution of Flake8:
The specified config file does not exist: foo.cfg
```
To me the traceback seems misleading, since it doesn't point to the actual call site. If that's acceptable then I can re-add the assertions, otherwise let me know if you have a better solution in mind.

Reply:
I think this should be

```python
    def exit_code(self) -> int:
        """Return the program exit code."""
        if self.catastrophic_failure:
            return 1
        assert self.options is not None
        if self.options.exit_zero:
            return 0
        else:
            return int(self.result_count > 0) 
```

eventually I want to factor away the `Application` class since the lateinit nullable attributes pose a lot of problems and we have to side-step the typechecker a lot

Reply:
actually it's slightly more involved -- here's a fix for that:

```diff
diff --git a/src/flake8/main/application.py b/src/flake8/main/application.py
index ee46f6a..1894881 100644
--- a/src/flake8/main/application.py
+++ b/src/flake8/main/application.py
@@ -115,10 +115,13 @@ class Application:
 
     def exit_code(self) -> int:
         """Return the program exit code."""
-        if self.options and self.options.exit_zero:
-            return int(self.catastrophic_failure)
+        if self.catastrophic_failure:
+            return 1
+        assert self.options is not None
+        if self.options.exit_zero:
+            return 0
         else:
-            return int((self.result_count > 0) or self.catastrophic_failure)
+            return int(self.result_count > 0)
 
     def find_plugins(
         self,
@@ -390,6 +393,7 @@ class Application:
         except exceptions.EarlyQuit:
             self.catastrophic_failure = True
             print("... stopped while processing files")
-
-        if self.options and self.options.count:
-            print(self.result_count)
+        else:
+            assert self.options is not None
+            if self.options.count:
+                print(self.result_count)
```

actually, I think this is broken on `main` so I'll split it out to a separate patch